src/components/KeyboardAwareScrollView/index.tsx

```diff
@@ -345,42 +586,63 @@ const KeyboardAwareScrollView = forwardRef<
     useAnimatedReaction(
       () => input.value,
       (current, previous) => {
+        // Handle input height changes while maintaining focus
         if (
           current?.target === previous?.target &&
           current?.layout.height !== previous?.layout.height
         ) {
           const prevLayout = layout.value;
 
           layout.value = input.value;
-          scrollPosition.value += maybeScroll(keyboardHeight.value, true);
+
+          const visibleRect =
+            scrollViewLayout.value.height - keyboardHeight.value;
+
+          // Scroll if input is taller than visible area and cursor is out of view
+          const absoluteY = layout.value?.layout.absoluteY || 0;
+          const visibleBottom = visibleRect - bottomOffset;
+          const cursorPosition = absoluteY + cursorOffset.value;
+
+          if (cursorPosition > visibleBottom - bottomOffset) {
+            log(
+              "Handle input height change",
+              cursorPosition,
+              ">",
+              visibleBottom,
+              "-",
+              bottomOffset,
+            );
+            maybeScroll(true);
+          }
+
           layout.value = prevLayout;
         }
       },
-      [],
+      [bottomOffset],
     );
 
     const view = useAnimatedStyle(
-      () =>
-        enabled
-          ? {
-              // animations become choppy when scrolling to the end of the `ScrollView` (when the last input is focused)
-              // this happens because the layout recalculates on every frame. To avoid this we slightly increase padding
-              // by `+1`. In this way we assure, that `scrollTo` will never scroll to the end, because it uses interpolation
-              // from 0 to `keyboardHeight`, and here our padding is `keyboardHeight + 1`. It allows us not to re-run layout
-              // re-calculation on every animation frame and it helps to achieve smooth animation.
-              // see: https://github.com/kirillzyusko/react-native-keyboard-controller/pull/342
-              paddingBottom: currentKeyboardFrameHeight.value + 1,
-            }
-          : {},
+      () => ({
+        paddingBottom: enabled ? currentKeyboardFrameHeight.value : 0,
+      }),
       [enabled],
     );
 
+    useAnimatedReaction(
+      () => scrollY.value,
+      (current) => {
+        scrollTo(scrollViewAnimatedRef, 0, current, false);
+      },
+    );
+
     return (
       <ScrollViewComponent
         ref={onRef}
         {...rest}
         scrollEventThrottle={16}
+        onContentSizeChange={onContentSizeChange}
         onLayout={onScrollViewLayout}
+        onScroll={scrollHandler}
```

Comment:
It's not very trivial to pass handler in such way. We have to forward that event to external handlers. And it's tricky, because those handlers can be worklet/plain-js handlers. I'll link a chain of PR here in chronological order: https://github.com/kirillzyusko/react-native-keyboard-controller/pull/339 https://github.com/kirillzyusko/react-native-keyboard-controller/pull/408 https://github.com/kirillzyusko/react-native-keyboard-controller/pull/490 https://github.com/kirillzyusko/react-native-keyboard-controller/pull/500